app/services/amr/n3rgy_energy_tariff_inserter.rb

```diff
@@ -0,0 +1,194 @@
+module Amr
+  #This version makes some assumptions about what tariffs are
+  #loaded. Specifically, that there is a single days tariff.
+  #This is what v1 of the n3rgy API now supports.
+  #
+  #As a result the class ignores the date ranges returned by
+  #the `N3rgyTariffs` class. It just relies on that to map
+  #the list of prices into a more compressed representation
+  #which consists of time of day ranges that can be used to
+  #check the energy tariff prices
+  class N3rgyEnergyTariffInserter
+    def initialize(meter:, start_date:, tariff:, import_log:)
+      @meter = meter
+      @start_date = start_date
+      @tariff = tariff
+      @import_log = import_log
+      @import_log_error_messages = []
+    end
+
+    def perform
+      check_unexpected_tariff_format?
+      energy_tariff = latest_energy_tariff
+
+      if reject_as_zero_standing_charges? || reject_as_zero_tariffs?
+        update_existing_tariff(energy_tariff)
+        return
+      end
+
+      if energy_tariff.nil? || tariff_changed?(energy_tariff)
+        create_new_tariff
+        update_existing_tariff(energy_tariff)
+      end
+    end
+
+    private
+
+    def latest_energy_tariff
+      @meter.energy_tariffs.dcc.where(end_date: nil).order(created_at: :desc).first
+    end
+
+    #Change means:
+    #  Different type of tariff
+    #  Different standing charge, but same type of tariff
+    #  Different prices, but same type of tariff
+    #  Different differential periods
+    def tariff_changed?(energy_tariff)
+      return true unless same_tariff_type?(energy_tariff)
+      return true unless same_standing_charge?(energy_tariff)
+      return true unless same_prices?(energy_tariff)
+      false
+    end
+
+    def same_tariff_type?(energy_tariff)
+      energy_tariff.flat_rate? && flat_rate?
+    end
+
+    def same_standing_charge?(energy_tariff)
+      energy_tariff.energy_tariff_charges.where(charge_type: :standing_charge, value: standing_charge, units: :day).any?
+    end
+
+    #Should already have checked if existing and new tariff are
+    #same type
+    def same_prices?(energy_tariff)
+      if energy_tariff.flat_rate?
+        return energy_tariff.energy_tariff_prices.first.value == rates.values.first
+      else
+        return false unless energy_tariff.energy_tariff_prices.count == rates.keys.count
+        rates.each do |times, price|
+          return false unless energy_tariff.energy_tariff_prices.where(start_time: times.first.to_s, end_time: times.last.to_s, value: price).any?
+        end
+      end
+      true
+    end
+
+    def flat_rate?
+      rates.keys.count == 1
+    end
+
+    def standing_charge
+      summary_tariff[:standing_charges].values.first
+    end
+
+    def reject_as_zero_standing_charges?
+      if standing_charge <= 0.0
+        log_error("Standing charge returned from n3rgy for #{@start_date} are zero #{standing_charge}")
+        return true
+      end
+      false
+    end
```

Comment:
maybe this could be rewritten to avoid the return? (and similar for `reject_as_zero_tariffs?` below)
```
    def reject_as_zero_standing_charges?
      if standing_charge <= 0.0
        log_error("Standing charge returned from n3rgy for #{@start_date} are zero #{standing_charge}")
        true
      else
        false
      end
    end
```

Reply:
or even simpler?
```
    def reject_as_zero_standing_charges?
      return false unless standing_charge <= 0.0

      log_error("Standing charge returned from n3rgy for #{@start_date} are zero #{standing_charge}")
      true
    end
```